core/store/src/trie/mem/flexible_data/encoding.rs

```diff
@@ -0,0 +1,132 @@
+/// the memory being decoded from.
+pub struct RawDecoder<'a> {
+    data: *const u8,
+    pos: usize,
+    _marker: std::marker::PhantomData<&'a ()>,
+}
+
+impl<'a> RawDecoder<'a> {
+    /// Starts decoding from the given memory location. The location should be
+    /// a pointer returned by RawEncoder::finish.
+    pub fn new(data: *const u8) -> RawDecoder<'a> {
+        RawDecoder { data, _marker: std::marker::PhantomData, pos: 0 }
+    }
+
+    unsafe fn ptr(&mut self) -> *const u8 {
+        self.data.offset(self.pos as isize)
+    }
+
+    /// Decodes a fixed-size field at the current decoder position, and then
+    /// advances the position by the size of the field.
+    pub unsafe fn decode<T>(&mut self) -> &'a T {
+        let result = &*(self.ptr() as *const T);
+        self.pos += std::mem::size_of::<T>();
+        result
+    }
+
+    /// Like decode, but returns the reference as mutable.
+    pub unsafe fn decode_as_mut<T>(&mut self) -> &'a mut T {
+        let result = &mut *(self.ptr() as *mut T);
+        self.pos += std::mem::size_of::<T>();
+        result
+    }
+
+    /// Decodes a fixed-sized field at the current position, but does not
+    /// advance the position.
+    pub unsafe fn peek<T>(&mut self) -> &'a T {
+        &*(self.ptr() as *const T)
+    }
+
+    /// Decodes a flexibly-sized part of the data at the current position,
+    /// and then advances the position by the size of the flexibly-sized part,
+    /// as returned by `header.flexible_data_length()`.
+    pub unsafe fn decode_flexible<T: FlexibleDataHeader>(&mut self, header: &'a T) -> T::View<'a> {
+        let length = header.flexible_data_length();
+        let view = header.decode_flexible_data(self.ptr());
+        self.pos += length;
+        view
+    }
+
+    /// Takes a fixed field from the current position, and then advances the
+    /// position by the size of the field. This is useful for dropping the fixed
+    /// field when the overall structure is being dropped.
+    pub unsafe fn take_fixed<T>(&mut self) -> T {
+        let field = &mut *(self.ptr() as *mut MaybeUninit<T>);
+        let result = std::mem::replace(field, MaybeUninit::uninit()).assume_init();
+        self.pos += std::mem::size_of::<T>();
+        result
+    }
+
+    /// Drops a flexibly-sized part of the data at the current position, and
+    /// then advances the position by the size of the flexibly-sized part.
+    pub unsafe fn drop_flexible<T: FlexibleDataHeader>(&mut self, header: &T) {
+        let length = header.flexible_data_length();
+        header.drop_flexible_data(self.ptr() as *mut u8);
+        self.pos += length;
+    }
+
+    /// At the end of decoding, converts the memory buffer to a Box so that the
+    /// memory itself can be deallocated. Requires that the decoding be complete
+    /// (so that we have the right size for the memory allocation).
+    pub unsafe fn take_encoded_data(&self) -> Box<[u8]> {
```

Comment:
Should give out `Box<[MaybeUninit<u8>]>`. Also, should take `self` and not `&self`, as there’s definitely no way this object will still be valid after running this.

Considering how this function means `RawDecoder` took ownership of the data, I’d suggest also making it take in `Box<[MaybeUninit<u8>]>` in `fn new()`. If it cannot, then this function is probably not something that should be here, as it assumes there is actually a `Box` that has been `into_raw`ed and can be recovered.